### ISSUE TITLE
[MS] Define a MSPileup pileup data structure to be persisted in the database

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#11424]
**Impact of the new feature**
New WMCore microservice

**Is your feature request related to a problem? Please describe.**
Creation of a new microservice to make pileup management more efficient and easier.
Subtask of https://github.com/dmwm/WMCore/issues/9779

**Describe the solution you'd like**
We need to define the relevant attributes for a pileup object - which will be persisted in the database - such that this future microservice can automatically act on it according to the values provided by an user. This needs to be further discussed with the P&R team.

The expected values are supposed to be extracted from the P&R campaign dictionary:
https://github.com/CMSCompOps/WmAgentScripts/blob/master/campaigns.json

A candidate data structure has been mentioned in this ticket: https://github.com/dmwm/WMCore/issues/9779. Here is an slightly improved version of it:
```
{"pileupName": string with the pileup dataset,
  "pileupType": string with a constant value (classic or premix),
  "insertTime": "human readable date/time of when this pileup object was added to the system",
  "lastUpdateTime": "human readable date/time of when this pileup object was last modified by an user",
  "expectedRSEs": ["Disk1", "Disk2", etc],  # these values need to be validated against Rucio
  "currentRSEs": ["Disk1", "Disk3"],  # values provided by the MS itself
  "fullReplicas": integer,  # total number of replicas to keep on Disk (will eventually make expectedRSEs meaningless)
  "campaigns": ["camp1", "camp2", etc]  # list of workflow campaigns using this pileup
  "containerFraction": real number with the container fraction to be distributed (TBFD)
  "replicationGrouping": string with a constant value (DATASET or ALL, to be in sync with Rucio)
  "activatedOn": "human readable date/time of when this pileup object was **last** activated,
  "deactivatedOn": "human readable date/time of when this pileup object was **last** deactivated,
  "active": True/False}
```

where:
* pileupType: string with either "premix" or "classic" value. Identifying which pileup data it corresponds to.
* insertTime: contains the timestamp of when this pileup id was first defined in the microservice (by an user)
* lastUpdateTime: contains the timestamp of the last modification made to this pileup structure (by an user)
* expectedRSEs: to be provided by someone (P&R and/or DM). This list of RSEs need to be properly validated against the known RSE names and ensured that it corresponds to a Disk RSE.
* currentRSEs: to be filled up by the microservice itself, as a result of continuous data location. Once there is a lock for the pileup id in question, the RSE name should be written here.
* fullReplicas: will eventually superseed "expectedRSEs". To be used whenever the microservice is mature enough and automatic data placement decisions can be performed.
* active: whether the pileup id is meant to be used by any campaign/workflow. `False` means it's no longer used and that replicas should be removed from Disk.

These pileup objects are supposed to be stored in a central database (MongoDB?) in a single document, with a list data structure like:
```
[<pileup object 1>, <pileup object 2>, ..., <pileup object n>]
```

UPDATE (on Jan/13/2023):
As discussed with Hasan in the P&R meeting, we agreed that these objects need to carry a campaign information in which they are actually being used. Here is a list of pileup parameters that we would like to add, either for an immediate use or for a potential near future:
* `campaigns`: a list of campaign strings. This information is directly associated to the campaign configuration

we also think that there would be value in having the following parameters, but their utilization isn't yet fully clear and it might depend on further discussions with other groups. Given that this object is supposed to be small enough, I think we should add them, even if their functionality isn't implemented at this moment. They are:
* `containerFraction`: a real number that corresponds to the fraction number of blocks to be replicated in a given RSE (default value is 1.0, full container at the same RSE - only supported case at this very moment)
* `replicationGrouping`: matching the grouping granularity field provided by Rucio. Allowed values would be DATASET (data placement by block) or ALL (whole container is placed under the same RSE)
* `activatedOn`: human friendly date of when this pileup last became active
* `deactivatedOn`: human friendly date of when this pileup last became deactive.

Alternative for the activatedOn/deactivatedOn parameters would be to provide an activeTransition list that gets a new item every time the pileup changes its active status.

**Describe alternatives you've considered**
Perhaps it would be important to plan for two other fields to the configuration, like:
* `pileupSize`: to be updated by the microservice itself with the current size of the pileup in bytes
* `rulesList`: list of strings (rules) used to lock the pileup id.

**Additional context**
None

**Comment by vkuznet** [comment]
First look at https://github.com/CMSCompOps/WmAgentScripts/blob/master/campaigns.json reveals usual pattern, this document is not static and has dynamic keys. My advise to avoid this in a future services and if we will create new pileup schema it should not have dynamic keys. The proposed data structure so far is static and I'm in favor of enforcing it.

**Comment by vkuznet** [comment]
@amaltaro , while I'm awaiting on https://github.com/dmwm/WMCore/pull/11430 and https://github.com/dmwm/WMCore/issues/11423 I start looking into this issue. Here I want to clarify few things:
- which database will be used for pileup data, relational (MariaDB) or document-oriented (MongoDB)?
- do we need data validation, I think we do
- what would be the input for data create, is it campaign.json or something else or class will create structure outlined in this ticket and we'll update it later
- what are the foreseen APIs, e.g.
   - `def create(self, cdict)` creates MSPileup data from given campaign.json dictionary, and stores it to persistent storage
   - `def update(self, data)` updates MSPileup data in persistent storage with given data dictionary
   - `def store(self, data)` stores MSPileup data to persistent storage, e.g. someone can provide pileup data to store
   - `def get(self, pileupId)` fetches MSPileup data from persistent storage for a given pileupID
   - `def delete(self, pileupId)` deletes MSPileup data in persistent storage for given pileupID

Any additional input is welcome.

**Comment by amaltaro** [comment]
> which database will be used for pileup data, relational (MariaDB) or document-oriented (MongoDB)?

I would rather not depend on a relational database in central services. So I'd suggest to stick to MongoDB.

> do we need data validation, I think we do

yes. I think most of the fields have to be validated for data type and, in some cases, for a short list of allowed values.

> what would be the input for data create, is it campaign.json or something else or class will create structure outlined in this ticket and we'll update it later

the input has to be a python dictionary with the data structure suggested in the beginning of this ticket (service-related parameters should not be required though)

Your list of internal APIs look good to me. I'd though suggest to use more meaningful names, e.g.: createPileupObj, updatePileupObj, etc

**Comment by amaltaro** [comment]
I missed your question about the initial pileup data to be inserted into the database. For that, we will need to consolidate on the data structure to be used within this microservice. Once we agree on that, we need to parse the Unified configuration:
https://github.com/CMSCompOps/WmAgentScripts/blob/master/campaigns.json

and create a json file out of it (with a list of dictionaries, where each dictionary contains one pileup configuration). This json file will then be used to initially populate the database.

About this script, we can discuss the details once you get your hands dirt, but the way to fill up this attribute `pileupType` will be by parsing the dataset name and checking whether the datatier corresponds to PREMIX or not. Allowed values for this one field should be "classic" or "premix".

**Comment by vkuznet** [comment]
Alan, thanks for initial details. It would be useful if you'll show how conversion suppose to work on real examples. For instance, here is an example with prefix data:
```
  "RunIIAutumn18DRPremix": {
    "SiteBlacklist": [
      "T2_CH_CERN",
      "T2_CH_CERN_HLT"
    ],
    "fractionpass": 0.95,
    "go": true,
    "lumisize": -1,
    "maxcopies": 1,
    "overflow": {
      "PREMIX": {}
    },
    "resize": "auto",
    "secondaries": {
      "/Neutrino_E-10_gun/RunIISummer17PrePremix-PUAutumn18_102X_upgrade2018_realistic_v15-v1/GEN-SIM-DIGI-RAW": {
          "SecondaryLocation": [
                    "T1_US_FNAL_Disk"
                        ]
      }
    },
    "secondary_AAA": true,
    "tune": true
  },
```
SO, if I'm correct it will be translated to:
```
{
"pileupName": "RunIIAutumn18DRPremix",
"pileupType": "classic", # here I have no idea how to assign classic or prefix, and need some guidance from your site
"active": true,
}
```
The time info is clear, the RSE info I expect will come from Rucio, but I need to understand should it be fetched during conversion from campaigns.json script or should be provided later. If you think these details is too complicated to explain, I suggest we host short meeting next week to discuss them.

**Comment by todor-ivanov** [comment]
Hi @vkuznet @amaltaro,

There are  few stuff here that need to be taken into account, and which I do not seem to notice being touched in any of the discussions so far.
* We need to consider  which is the target/subject of this data structure - it is neither the Campaign nor the workflow, but it is rather the pilep itself. The document with the Campaign configuration is one thing, but it relates to the campaign, which may or may not define a pileup. And a campaign may or may not have any workflow currently active in the system, but still to be present in the Camapaign configuration .json file. So a constant iteration through this document is going to be needed here during the service runtime and updating of the database based on what is currently active in the system.
* Then, the granularity, is also something that needs to be stressed - I suppose we will have to upload a single document per PU, rather than a huge one including all possible pileups - similar to the campaign configuration json file.
* Then, IIRC, a PU may be configured not only on the Campaign level, but also in the workflow description. And going even further down, it may be also:
    * At WF's  top level configuration  
    * Or at the WF's per Task/Step configuration

All that said, I'd suggest, instead of only going through the campaign document and converting it to a single huge PU describing document, we go and create one document per active PU and then we add two separate fields in the data structure, to keep track of which campaigns and/or workflows are referring to this PU. 
Something like:
```
{
...,
'campaigns': {'name': "",
                         'active': True|False}
'workflows': {'name':""
                        'status' : new|accuired|open-running..}
}
```

**Comment by amaltaro** [comment]
After discussion in the P&R meeting today, I updated the initial description with what was discussed with Hasan today (check out the UPDATE part).

I also have some comments on what Todor said above:

> We need to consider which is the target/subject of this data structure - it is neither the Campaign nor the workflow, but it is rather the pilep itself. The document with the Campaign configuration is one thing, but it relates to the campaign, which may or may not define a pileup. And a campaign may or may not have any workflow currently active in the system, but still to be present in the Camapaign configuration .json file. So a constant iteration through this document is going to be needed here during the service runtime and updating of the database based on what is currently active in the system.

instead of a polling system, I think the best would be to have data being pushed. So, whenever changes happen to a campaign, that the relevant pileup information to be propagated to MSPileup.

> Then, the granularity, is also something that needs to be stressed - I suppose we will have to upload a single document per PU, rather than a huge one including all possible pileups - similar to the campaign configuration json file.

Yes, I think this will be a better approach than what I initially thought (single document for all these configs). With multiple documents, we can use the MongoDB query language to optimize data look up.

> Then, IIRC, a PU may be configured not only on the Campaign level, but also in the workflow description. And going even further down, it may be also:
>    At WF's top level configuration
>    Or at the WF's per Task/Step configuration

this will increase the complexity of the system and I don't think tracking this in a workflow basis gives us much benefit. In other words, I don't think we should inspect workflow descriptions to see which pileup they are using (other than during assignment and data placement (staging)).

Keeping workflow names and their statuses is another source of inconsistency as well. So we better keep it simpler. This is likely a relevant information though, and we should discuss how the best way to monitor it (likely outside of MSPileup though).

**Comment by vkuznet** [comment]
@amaltaro , @todor-ivanov , even though I appreciate your textual feedback it does not clarify the things about implementation. Here are few questions which still require clarification from your side (I suggest to clarify them during our Tue meeting):
- who suppose to parse campaign.json, does it MSPileup job which should read it from somewhere (if so what is a data source or service which provides it), or somebody will provide (push) this document to MSPileup (in this case do we need API which will expect this campaign.json)?
   - In other words, do we need API for accepting campaigns.json or some process upstream will parse it and will call `createPileupObj(doc)` API with required document
- please clarify how new pileup document is created, is it through API, e.g. `createPileupObj(doc)` which will accept data-structure we have in description, or by any other means (if so how). In other words, we'll have POST HTTP end-point which will be used by data-ops to pass pileup JSON doc (with data-structure we discuss here) and data layer will create it (and fill necessary fields, like timestamp). If this is the case then `createPileupObj` API will only validate given doc, fill out necessary fields and store it to database, right?
   - If so, which fields of data-structure are mandatory and which are optional?
   - It is still unclear if we'll be given campaigns.json document how to determine some attribute values, like pileupType, campaign attributes (which are not part of campaigns.json)?
   - what is the order and update policy of creation of pileup docs, does parsing campaigns.json go first or we will be given pileup data-structure, in this case do you allow update one or another (upsert MongoDB operation)?

**Comment by amaltaro** [comment]
As discussed in the chat with Todor, Valentin and (partially) Hasan this morning. Here is a summary of the discussion around the CRUD operations and how things should behave:
1. document creation (POST)
 * if document already exists, return an error saying that the pileup already exist (http error code)
   * returns a list of dictionaries
 * pileupName is unique key
 * campaigns.json is meant to be used ONE TIME, only for the initial translation and database population
 * maybe validate pileupName by calling Rucio
 * must enforce some CRIC-based authz

2. document deletion (DELETE) 
 * takes a pileupName as input and deletes it
   * if pileupName document does not exist, return an specific error
 * must enforce some CRIC-based authz

3. document retrieval (GET)
 * get a pileup config with the pileupName
   * if pileupName document does not exist, return an specific error
 * filters (query string):
   * (get all doc names) retrieve a list with all pileupName ids
   * (get all full docs) retrieve a list with the full documentation of all the pileups
   * given a string campaign name, retrieve all pileup names used by this
   * BONUS: accept expressions/queries that can be executed on the database
 * returns a list of dictionaries [{pileupName: abc}, {pileupName: ddd}, ...]

4. document update (PUT)
 * schema must be enforced before writing to the database
 * document in the database will be **updated** (field value(s) are replaced!) by the data provided by the client
 * given a pileupName as key, and other parameters to be updated, update one document in the database
 * update: key'ed by the pileupName, carrying dictionary with data to be updated (dict cannot have pileupName)
 * returns an empty list (http code 200 or something). If failed, return error code in a list of dict

PS.: I am pasting this same information to #11425 issue as well.

**Comment by amaltaro** [comment]
@vkuznet Valentin, after a brief discussion about MSPileup with the P&R team today, @haozturk expressed the following concerns/questions:
1) what are the mandatory and optional parameters?
2) what are the default values?
3) and ill RSE validation (which is expected AFAICT, given that we still do not pass the RSEs available in Rucio).

Based on that, I revisited what was documented in the wiki:
https://github.com/dmwm/WMCore/wiki/ReqMgr2-MicroService-Pileup#data-structure

and properly categorized what is a) mandatory; b) optional; and c) filled by the service. I would be in favor of reopening this ticket to address the issues above, but feel free to open a new one and link it to this one, if you prefer.

PS.: Note that the wiki has been updated to the `pileup` endpoint and it also contains the `ruleIds` parameter now.
